### DISCUSSION TITLE
Troubleshooting frequently made mistakes

### DISCUSSION THREAD

**Discussion by Lukas-Kunz** [ZEN-universe/ZEN-garden#298] [General ZEN-garden]
Are you facing any problems while running your dataset that you cannot resolve? Have a look at the following Q&A:

**Q: While running the framework I encounter a _ModuleNotFoundError_.**
**A: Ensure that you have activated the newest version of the ZEN garden virtual environment** (All the needed modules are included there). To do so, go to settings (file->settings) and check if the default python interpreter is the ZEN garden environment. If that is the case, check if the missing module is contained in the package list of the environment (screenshot):
![image](https://github.com/RRE-ETH/ZEN-garden/assets/114185605/8a699804-87f6-4e22-a684-81cf2fb2f9e9)
Now there are two options:
a. Module is contained: Most certainly you are running the framework with a configuration that specifies a different python interpreter. Check if the default project python interpreter (which should be the ZEN garden environment) is chosen in the configuration you are using to execute the framework:
![image](https://github.com/RRE-ETH/ZEN-garden/assets/114185605/9969d96d-4fcc-4df6-bd9c-4446cf8ce00c)

b. You are not using the newest ZEN garden environment version. To re-install the environment, run the following two commands in your Anaconda Prompt (must be  done in the directory, where the ZEN garden package is located-> use command `cd/ZEN-garden` (path of ZEN garden)):
`conda remove --name zen-garden-linopy --all` (replace zen-garden-linopy by your environment name if it is named differently)
`conda env create -f zen_garden_env.yml`
After updating the ZEN garden environment, ensure again if the environment is set as the default python interpreter and if the module can now be found in the package list.

**Q: My dataset is infeasible.**
**A: Estimate if the following criteria are fulfilled:**

1. Can the specified carrier demands be produced with the available technologies?
2. Can the specified demands be met without violating the defined carbon emissions specifications?
3. Can the needed technologies be built on time? (construction_time > 0 and capacity_existing = 0)

**Q: When I try to run my dataset I get the error _FileNotFoundError: system.py not found in dataset: C:\ZEN-garden\data\my_dataset_.**
**A: Your dataset is not located in the _data_ folder.** To execute a dataset it must be placed in the _data_ directory.

**Q: I get the error "pint.errors.UndefinedUnitError: 'KWh' is not defined in the unit registry", what's wrong?**
**A: Your prefix should/shouldn't be capitalized e.g. kilo --> k (not K), mega --> M (not m), etc.**

**Q: I encounter the error "pandas.errors.ParserError: Error tokenizing data. C error: Expected 3 fields in line 10, saw 4" while reading the input data.**
**A: There's an input csv file showing more commas (separators) than the other rows do.**

**Comment by hegnerl** [comment]
**Q: when submitting a job to euler using _sbatch_, I'm getting the error** _sbatch: error: Invalid --distribution specification_ **despite not passing this argument at all or passing it correctly into sbatch. Or I get another error related to an input argument that looks to be formulated correctly.**
**A: make sure the input command is being parsed correctly.** 
Copy-pasting a command such as _sbatch --ntasks=32 --mem-per-cpu=2048 --time=2:00:00 --ntasks-per-node=32 --wrap=“python -m zen_garden”_ from e.g. slack could result in a wrong interpretation of the arguments passed. --> the quotation marks (") might not be parsed as such. Manually typing out the command will fix this issue


**Q: when setting up euler, when building the virtual environment (./create_env_euler.sh), I'm getting the error:** 
	"Using cached PyYAML-5.4.1.tar.gz (175 kB)
	Installing build dependencies ... done
	Getting requirements to build wheel ... error 
	error: subprocess-exited-with-error
	... 
	AttributeError: cython_sources
	[end of output]"
**A:** 1) add the 2 highlighted lines to _create_env_euler.sh_. 
_pip install "cython<3.0.0" wheel
pip install "pyyaml==5.4.1" --no-build-isolation_
2) make sure that the gurobipy version in _requirements.txt_ is specified to be ==10.0 (this might change in future)

<img width="1389" alt="Screenshot 2023-12-13 at 18 10 17" src="https://github.com/ZEN-universe/ZEN-garden/assets/86005408/e1a6f733-d8b4-43f2-bf0f-b3e2140be8ba">
<img width="1238" alt="Screenshot 2023-12-13 at 18 09 58" src="https://github.com/ZEN-universe/ZEN-garden/assets/86005408/d0bc2583-3e5b-4a4d-8f99-5c5dd35fdd2c">
<img width="623" alt="Screenshot 2023-12-13 at 18 05 41" src="https://github.com/ZEN-universe/ZEN-garden/assets/86005408/988a0086-6e2d-4bcb-8cc0-e55fabe43b50">
<img width="263" alt="Screenshot 2023-12-13 at 18 29 16" src="https://github.com/ZEN-universe/ZEN-garden/assets/86005408/b3f86f03-8c4a-46c9-bbef-3c042dfbac17">
